### ISSUE TITLE
Table view cleanup (currently it is completely broken)

### ISSUE THREAD

**Issue by esjewett** [humanitiesplusdesign/fibra#14]
Several issues (please add more):

- [ ] How do we handle multiple types? Probably for starters we just put the primary item type in the table
- [ ] Display locally stored properties only
- [ ] Download capability?

**Comment by jiemakel** [comment]
Regarding the local vs remote properties, we still currently have the problem of not actually having any local properties :). 

So, to test that, we need to either develop the functionality to select properties to import OR, once Nicole defines one of the "working sets", I can just import the stuff relevant to that outside of the UI.

**Comment by esjewett** [comment]
Well, that is a fair comment :-) I guess we'll need some way to deal with that for the time being.

**Comment by alexsherman** [comment]
I'm currently struggling with this. I'm completely new to both Pug and Angular, and I'm a little confused by the item api (and if there are actually local properties yet). I've been trying to do stuff with ng-if and ng-switch to make separate tables based on the type, but I haven't had success, and I'm confused about how to find the type for any given item--as I've been messing around, it seems like the type of the item is just randomly found somewhere in item.Localproperties. 

I feel like this should be easier than I've been making it, but I can't seem to get Angular to do what I want. Any help would be appreciated.

**Comment by esjewett** [comment]
Yes, it's very confusing, both on the angular front and on the item API front.

At the moment, in the explore component (which is where the table is, in explore.pug), the `$ctrl` has `$ctrl.primaryItems`, `$ctrl.secondaryItems`, and `$ctrl.tertiaryItems` (which I think is always blank currently). Those are items that are already sorted out by type. The types they are sorted by are in order on `$ctrl.orderedTypes`. Right now these types are hard-coded, but they'll be read off of the configuration shortly, I think. So, that's where your items come from.

You can look at the `filterItemsByType` in the `ExploreComponentController` to get an idea of where the types live. There are not currently many (if any) local properties on newly added nodes, but I think type should be on there, and many of the nodes have quite a few localProperties. I'm not clear on if this is because the local properties aren't really local at the moment, or if these are old data. In any case, it is something to work with :-)

<img width="902" alt="screen shot 2016-11-03 at 9 18 37 pm" src="https://cloud.githubusercontent.com/assets/4822/19992355/1f811afa-a20b-11e6-9d1d-549a063bbafb.png">

Above is a screenshot of a couple of the local properties of a single item. Notable things:
 * The `PropertyToValues` object has a field `value`. This is the type of property. In our example, the values are `http://www.w3.org/1999/02/22-rdf-syntax-ns#type` and `http://www.w3.org/2004/02/skos/core#altLabel`. Eetu can speak to this better than I, but these define the semantics of the properties as the "type" and the "alternative label".
* Properties on different items that have the same `PropertiesToValues.value` value will go in the same column of the table.
* The label of this value, and likely the header of the table column, will be the `label.value` at the same level of the object. So, in this case, `"type"` and `"alt Label"`.
* The actual property value associated with this property is found under the `values` array, in the `SourcedNodePlusLabel` objects (there can be many). You'll probably want to display `SourcedNodePlusLabel.label.value` in the table cell (perhaps multiple values separated in some way). In this case the label of the `"type"` property will be `"Person"` and the first label of the `"alt Label"` property will be `"philosopher"`.

I would suggest writing small helper functions on the controller to transform the data structure to something a little easier to tabulate. :-)

As far as Angular tables, the secret is `ng-repeat`. I think start with something like this and then start getting fancier, plug it in to the controller, etc.:

```html
<table>
  <tr ng-repeat="row in [[1,2,3],[4,5,6],[7,8,9]]">
    <td ng-repeat="cell in row">
      {{cell}}
    </td>
  </tr>
</table>
```

In .pug, that's

```pug
table
  tr(ng-repeat="row in [[1,2,3],[4,5,6],[7,8,9]]")
    td(ng-repeat="cell in row") {{cell}}

```
Simple example to play with is here, but feel free to shout if you have problems: https://jsfiddle.net/karL6hn8/

**Comment by jiemakel** [comment]
Just to repeat: 

Currently we have items, which have multiple properties, which have multiple values. All of these derive from decorated RDF nodes (an RDF node on the other hand is either a resource with an URI as the `value` or a literal with the literal as `value`). The primary decoration is that they have labels, which are undecorated literal nodes (the reason the labels are nodes instead of just plain text [as they initially were] is that nodes retain language information, which may be important at a later point). 

Back to the items: for UI purposes, the properties are classified as either local or remote, depending on if they originate from our local store or from the configured remote authorities (that is why they are `SourcedNode`s, which also denote the source). 

To see how these can be processed in Angular, you can also look at `sparql-item.pug` (and Pug, formerly Jade, is just an alternate syntax for HTML, you can quickly get accustomed to it for example by experimenting at http://html2jade.org/)

However! Currently in construct, all remote properties are hacked to show up in `localProperties`. This is because at present, our local store actually only contains the absolute minimum of information on each item (the label, type and equivalencies with remote sources) , which would result in a construct graph with no links.

I'll dismantle this hack once I've pulled a sensible set of test properties to the local store.

Now back to types and labels. As Ethan said, as far as the RDF data model is concerned, types and labels are just certain properties with a commonly agreed-upon meaning (here, rdf:type, skos:altLabel and skos:prefLabel). However, for labels, I currently also have logic to transfer a "best" label to the `label` javascript property, so every time we need to label something we don't have to iterate through all the various label properties to discover which one of them was best. For types, this hasn't been implemented yet, but I think I should also do a `primaryType` js property, as we're probably going to need to access the type in equally many places

**Comment by alexsherman** [comment]
Awesome, thank you both for the information. Both of these posts cleared a lot of stuff up for me. I will try to push soon with better looking / more functional tables.
